### PR TITLE
Cache the results of a WalkMapper

### PR DESCRIPTION
**Pull request by kaushikcfd**
Array expressions are a DAG, if an expression is traversed do not re-traverse it.

### CODE REVIEW

File: pytato/transform.py [hunk 1]
```diff
@@ -259,6 +259,20 @@ class WalkMapper(Mapper):
     .. automethod:: visit
     .. automethod:: post_visit
     """
+    def __init__(self) -> None:
```
Comment (inducer) [inline pytato/transform.py:262]: I think not every `WalkMapper` will want this. Make this a subclass?

File: pytato/transform.py [hunk 2]
```diff
@@ -259,6 +259,20 @@ class WalkMapper(Mapper):
...
+        self._cache: Dict[ArrayOrNames, Any] = {}
```
Comment (inducer) [inline pytato/transform.py:263]: I don't think `WalkMapper`s return anything... this can probably be a set?

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: pytato/transform.py (hunk 1 above)
Comment (kaushikcfd) [reply pytato/transform.py]: Yes, thanks. There is a `CachedWalkMapper` now.

File: pytato/transform.py (hunk 2 above)
Comment (kaushikcfd) [reply pytato/transform.py]: Done. fc06fc2

**Comment by inducer** [comment]
Thanks!

**Pull request merged**